File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    at_most = min(times) * n
    
    answer = at_most
    
    # 심사 시간이 가장 짧은 시간이 모두를 검사하는 시간을 최댓값으로 본다.
    l, r = 1, at_most

    while l <= r:
        mid = (l + r) // 2
        
        total_checked = sum([mid//time for time in times])

        if total_checked < n:
            l = mid + 1
        else:
            answer = min(answer, mid)
            r = mid - 1
            
    return answer

File: 프로그래머스/3/43238. 입국심사/입국심사.java
class Solution {
    public long solution(int n, int[] times) {
        long maximumTime = times[0];
        for (int time: times) {
            maximumTime = Math.max(maximumTime, time);
        }
        
        long right = n * maximumTime;
        long left = 1;
        long answer = right;
        
        while (left <= right) {
            long count = 0;
            long mid = left + (right - left) / 2;
            
            for (int time: times) {
                count += mid / time;
                
                if (count >= n) break;
            }
            
            if (count >= n) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        return answer;
    }
}